stage1/usr_from_kvm/kernel/patches/401-plkvm.patch

```diff
@@ -0,0 +1,495 @@
++		flush_dcache_page(page);
++	} else {
++		/*
++		 * FIXME: Need more involved flushing to ensure that writes to
++		 * NVDIMMs are actually durable before returning.
++		 */
++		flush_dcache_page(page);
++		copy_to_plkvm(plkvm, mem + off, sector, len);
++	}
++
++	kunmap_atomic(mem);
++}
++
++static void plkvm_make_request(struct request_queue *q, struct bio *bio)
++{
++	struct block_device *bdev = bio->bi_bdev;
++	struct plkvm_device *plkvm = bdev->bd_disk->private_data;
++	int rw;
++	struct bio_vec bvec;
++	sector_t sector;
++	struct bvec_iter iter;
++	int err = 0;
++
++	sector = bio->bi_iter.bi_sector;
++	if (bio_end_sector(bio) > get_capacity(bdev->bd_disk)) {
++		err = -EIO;
++		goto out;
++	}
++
++	BUG_ON(bio->bi_rw & REQ_DISCARD);
++
++	rw = bio_rw(bio);
++	if (rw == READA)
++		rw = READ;
++
++	bio_for_each_segment(bvec, bio, iter) {
++		unsigned int len = bvec.bv_len;
++
++		BUG_ON(len > PAGE_SIZE);
++		plkvm_do_bvec(plkvm, bvec.bv_page, len,
++			    bvec.bv_offset, rw, sector);
++		sector += len >> SECTOR_SHIFT;
++	}
++
++out:
++	bio->bi_error = err;
++	bio_endio(bio);
++}
++
++static int plkvm_rw_page(struct block_device *bdev, sector_t sector,
++		       struct page *page, int rw)
++{
++	struct plkvm_device *plkvm = bdev->bd_disk->private_data;
++
++	plkvm_do_bvec(plkvm, page, PAGE_CACHE_SIZE, 0, rw, sector);
++	page_endio(page, rw & WRITE, 0);
++	return 0;
++}
++
++static long plkvm_direct_access(struct block_device *bdev, sector_t sector,
++			      void **kaddr, unsigned long *pfn)
++{
++	struct plkvm_device *plkvm = bdev->bd_disk->private_data;
++
++	if (!plkvm)
++		return -ENODEV;
++
++	*kaddr = plkvm_lookup_pg_addr(plkvm, sector);
++	*pfn = plkvm_lookup_pfn(plkvm, sector);
++
++	return plkvm->size - (sector * 512);
++}
++
++static const struct block_device_operations plkvm_fops = {
++	.owner =		THIS_MODULE,
++	.rw_page =		plkvm_rw_page,
++	.direct_access =	plkvm_direct_access,
++	.getgeo =		plkvm_getgeo,
++};
++
++/* Kernel module stuff */
++
++static int plkvm_count = 1;
++module_param(plkvm_count, int, S_IRUGO);
++MODULE_PARM_DESC(plkvm_count, "Number of plkvm devices to evenly split allocated space");
++
++static LIST_HEAD(plkvm_devices);
++static int plkvm_major = -1;
++
++/* plkvm->phys_addr and plkvm->size need to be set.
++ * Will then set virt_addr if successful.
++ */
++int plkvm_maplkvm(struct plkvm_device *plkvm)
++{
++	struct resource *res_mem;
++	int err;
++
++	res_mem = request_mem_region_exclusive(plkvm->phys_addr, plkvm->size,
++					       "plkvm");
++	if (!res_mem) {
++		pr_warn("plkvm: request_mem_region_exclusive phys=0x%llx size=0x%zx failed\n",
++			   plkvm->phys_addr, plkvm->size);
++		pr_err("Map fail\n");
++		while(1);
++		return -EINVAL;
++	}
++
++	plkvm->virt_addr = ioremap_cache(plkvm->phys_addr, plkvm->size);
++	if (unlikely(!plkvm->virt_addr)) {
++		err = -ENXIO;
++		pr_err("Map fail\n");
++		while(1);
++		goto out_release;
++	}
++	return 0;
++
++out_release:
```

Comment:
This debug code is probably not really useful for users, right?

And "unlikely" is basically a hint for the compiler telling that the condition is almost always false, so the compiler can pull up some more optimization tricks out of its sleeve.